cupy/cuda/pinned_memory.pyx

```diff
@@ -0,0 +1,270 @@
+        p[0] = <void*>self.ptr
+        return self.size()
+
+    def __getwritebuffer__(self, Py_ssize_t idx, void **p):
+        if idx != 0:
+            raise SystemError("accessing non-existent buffer segment")
+        p[0] = <void*>self.ptr
+        return self.size()
+
+
+cpdef PinnedMemoryPointer _malloc(Py_ssize_t size):
+    mem = PinnedMemory(size, runtime.hostAllocPortable)
+    return PinnedMemoryPointer(mem, 0)
+
+
+cdef object _current_allocator = _malloc
+
+
+cpdef PinnedMemoryPointer alloc_pinned_memory(Py_ssize_t size):
+    """Calls the current allocator.
+
+    Use :func:`~cupy.cuda.set_pinned_memory_allocator` to change the current
+    allocator.
+
+    Args:
+        size (int): Size of the memory allocation.
+
+    Returns:
+        ~cupy.cuda.PinnedMemoryPointer: Pointer to the allocated buffer.
+
+    """
+    return _current_allocator(size)
+
+
+cpdef set_pinned_memory_allocator(allocator=_malloc):
+    """Sets the current allocator.
+
+    Args:
+        allocator (function): CuPy pinned memory allocator. It must have the
+            same interface as the :func:`cupy.cuda.alloc_alloc_pinned_memory`
+            function, which takes the buffer size as an argument and returns
+            the device buffer of that size.
+
+    """
+    global _current_allocator
+    _current_allocator = allocator
+
+
+cdef class PooledPinnedMemory(PinnedMemory):
+
+    """Memory allocation for a memory pool.
+
+    The instance of this class is created by memory pool allocator, so user
+    should not instantiate it by hand.
+
+    """
+
+    def __init__(self, PinnedMemory mem, pool):
+        self.ptr = mem.ptr
+        self.size = mem.size
+        self.pool = pool
+
+    def __dealloc__(self):
+        if self.ptr != 0:
+            self.free()
+
+    cpdef free(self):
+        """Frees the memory buffer and returns it to the memory pool.
+
+        This function actually does not free the buffer. It just returns the
+        buffer to the memory pool for reuse.
+
+        """
+        pool = self.pool()
+        if pool and self.ptr != 0:
+            pool.free(self.ptr, self.size)
+        self.ptr = 0
+        self.size = 0
+
+
+cdef class PinnedMemoryPool:
+
+    """Memory pool on the machine.
+
+    A memory pool preserves any allocations even if they are freed by the user.
+    Freed memory buffers are held by the memory pool as *free blocks*, and they
+    are reused for further memory allocations of the same sizes.
+
+    Args:
+        allocator (function): The base CuPy pinned memory allocator. It is
+            used for allocating new blocks when the blocks of the required
+            size are all in use.
+
+    """
+
+    def __init__(self, allocator=_malloc):
+        self._in_use = {}
+        self._free = collections.defaultdict(list)
+        self._alloc = allocator
+        self._weakref = weakref.ref(self)
+        self._allocation_unit_size = 512
```

Comment:
OK I understand what you say. If the memory pool fails to allocate, it calls ``free_all_blocks`` and release unused memory.